lib/block_view.go

```diff
@@ -2295,6 +2480,272 @@ func (bav *UtxoView) _disconnectCreatorCoinTransfer(
+	if utxoOpsForTxn[operationIndex].Type != OperationTypeAcceptNFTBid {
+		return fmt.Errorf("_disconnectAcceptNFTBid: Trying to revert "+
+			"OperationTypeAcceptNFTBid but found type %v",
+			utxoOpsForTxn[operationIndex].Type)
+	}
+	txMeta := currentTxn.TxnMeta.(*AcceptNFTBidMetadata)
+	operationData := utxoOpsForTxn[operationIndex]
+	operationIndex--
+
+	// In order to disconnect an accepted bid, we need to do the following:
+	// 	(1) Revert the NFT entry to the previous one with the previous owner.
+	//  (2) Add back all of the bids that were deleted.
+	//  (3) Disconnect payment UTXOs.
+	//  (4) Unspend bidder UTXOs.
+	//  (5) Revert profileEntry to undo royalties added to BitCloutLockedNanos.
+	//  (6) Revert the postEntry since NumNFTCopiesForSale was decremented.
+
+	// (1) Set the old NFT entry.
+	if operationData.PrevNFTEntry == nil || operationData.PrevNFTEntry.isDeleted {
+		return fmt.Errorf("_disconnectAcceptNFTBid: prev NFT entry doesn't exist; " +
+			"this should never happen")
+	}
+
+	prevNFTEntry := operationData.PrevNFTEntry
+	bav._setNFTEntryMappings(prevNFTEntry)
+
+	// Revert the accepted NFT bid history mappings
+	bav._setAcceptNFTBidHistoryMappings(MakeNFTKey(prevNFTEntry.NFTPostHash, prevNFTEntry.SerialNumber), operationData.PrevAcceptedNFTBidEntries)
+
+	// (2) Set the old bids.
+	if operationData.DeletedNFTBidEntries == nil || len(operationData.DeletedNFTBidEntries) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: DeletedNFTBidEntries doesn't exist; " +
+			"this should never happen")
+	}
+
+	for _, nftBid := range operationData.DeletedNFTBidEntries {
+		bav._setNFTBidEntryMappings(nftBid)
+	}
+
+	// (3) Revert payments made from accepting the NFT bids.
+	if operationData.NFTPaymentUtxoKeys == nil || len(operationData.NFTPaymentUtxoKeys) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: NFTPaymentUtxoKeys was nil; " +
+			"this should never happen")
+	}
+	for _, paymentUtxoKey := range operationData.NFTPaymentUtxoKeys {
+		if err := bav._unAddUtxo(paymentUtxoKey); err != nil {
+			return errors.Wrapf(err, "_disconnectAcceptNFTBid: Problem unAdding utxo %v: ", paymentUtxoKey)
+		}
+	}
+
+	// (4) Revert spent bidder UTXOs.
+	if operationData.NFTSpentUtxoEntries == nil || len(operationData.NFTSpentUtxoEntries) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: NFTSpentUtxoEntries was nil; " +
+			"this should never happen")
+	}
+	for _, spentUtxoEntry := range operationData.NFTSpentUtxoEntries {
+		if err := bav._unSpendUtxo(spentUtxoEntry); err != nil {
+			return errors.Wrapf(err, "_disconnectAcceptNFTBid: Problem unSpending utxo %v: ", spentUtxoEntry)
+		}
+	}
+
+	// (5) Revert the creator's CoinEntry if a previous one exists.
+	if operationData.PrevCoinEntry != nil {
+		nftPostEntry := bav.GetPostEntryForPostHash(operationData.PrevNFTEntry.NFTPostHash)
+		// We have to get the post entry first so that we have the poster's pub key.
+		if nftPostEntry == nil || nftPostEntry.isDeleted {
```

Comment:
Sigh... no action needed here but this is going to be tricky to fix when we move posts to pkid...